### PR TITLE
Study view - density plot for mutation count vs fga

### PR DESCRIPTION
**Pull request by adamabeshouse**

### CODE REVIEW

File: src/pages/studyView/charts/scatterPlot/StudyViewDensityScatterPlot.tsx [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+const NUM_AXIS_TICKS = 8;
+const DOMAIN_PADDING = 15;
+const BIN_THRESHOLD = 0;
+
+@observer
+export default class StudyViewDensityScatterPlot extends React.Component<IStudyViewDensityScatterPlotProps, {}> implements AbstractChart {
+    @observable tooltipModel:any|null = null;
+    @observable pointHovered:boolean = false;
+    @observable mouseIsDown:boolean = false;
+    @observable shiftPressed:boolean = false;
```
Comment (zhx828) [inline src/pages/studyView/charts/scatterPlot/StudyViewDensityScatterPlot.tsx]: When keepCurrent is set to true, the shiftPressed event should not be used

File: src/pages/studyView/charts/scatterPlot/StudyViewDensityScatterPlot.tsx [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+        const minX = dataDomain.x[0];
+        const yRange = dataDomain.y[1] - dataDomain.y[0];
+        const minY = dataDomain.y[0];
+        return {
+            x: (val:number)=>(val/this.props.width)*xRange + minX,
+            y: (val:number)=>(val/this.props.height)*yRange + minY
+        };
+    }
+
+    @computed get dataSpaceToPixelSpace() {
```
Comment (zhx828) [inline src/pages/studyView/charts/scatterPlot/StudyViewDensityScatterPlot.tsx]: The downsampling method should be moved to util function and add tests cover it.

File: src/pages/studyView/charts/scatterPlot/StudyViewDensityScatterPlot.tsx [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+                    style={{
+                        data: {
+                            fill: color,
+                            stroke: "black",
+                            strokeWidth: 1,
+                            strokeOpacity: 0
+                        }
+                    }}
+                    size={3}
+                    symbol={this.binningData ? "circle" : "circle"}
```
Comment (zhx828) [inline src/pages/studyView/charts/scatterPlot/StudyViewDensityScatterPlot.tsx]: No need to do condition anymore

File: src/pages/studyView/charts/scatterPlot/StudyViewScatterPlot.tsx [hunk 4]
```diff
@@ -197,39 +197,62 @@ export default class StudyViewScatterPlot extends React.Component<IStudyViewScat
...
 
     @computed get data() {
-        return getDownsampledData(this.props.data, this.props.sampleToAnalysisGroup, this.dataSpaceToPixelSpace);
+        return getDownsampledData(this.props.data, this.dataSpaceToPixelSpace, this.props.sampleToAnalysisGroup);
     }
 
     @computed get scatters() {
         // sort NA to the beginning - it should be rendered at the bottom
         // otherwise, order doesnt matter. here we'll just use the same order as given in the prop, bc _.sortBy is stable sort
+        console.log("------------");
```
Comment (zhx828) [inline src/pages/studyView/charts/scatterPlot/StudyViewScatterPlot.tsx]: I think all code changes related to StudyViewScatterPlot are no longer needed.

Comment (adamabeshouse) [reply src/pages/studyView/charts/scatterPlot/StudyViewScatterPlot.tsx]: its necessary for compatibility, and theres some small changes here and there, I think just lets let these changes through

File: src/pages/studyView/charts/scatterPlot/StudyViewDensityScatterPlot.tsx (hunk 1 above)
Comment (adamabeshouse) [reply src/pages/studyView/charts/scatterPlot/StudyViewDensityScatterPlot.tsx]: shiftPressed determines whether keepCurrent is used. I'm not sure what you're saying

**Comment by adamabeshouse** [comment]
@zhx828 the dark grey background is intentional to provide contrast with the heatmap colors - the brighter colors in the yellow region are hard to see against white

**Comment by adamabeshouse** [comment]
after talking with @zhx828 and @jjgao we decided:
(1) x-axis always [0,1], y axis always [0, max in selection]
(2) white background, use 3/4 of current color range to avoid light colors at top
(3) legend along right hand side

and still only show data from selection

File: src/pages/studyView/charts/scatterPlot/StudyViewDensityScatterPlot.tsx (hunk 1 above)
Comment (zhx828) [reply src/pages/studyView/charts/scatterPlot/StudyViewDensityScatterPlot.tsx]: NM. I misunderstood the meaning of the shiftclick.

**Comment by zhx828** [comment]
@adamabeshouse it looks great, just some minor issues I noticed while testing
- [ ] filtered samples number on the header is not updated after brushing
- [ ] I filled an issue(https://github.com/cBioPortal/cbioportal/issues/5166) for the endpoint, you may see that while testing
- [ ] The filter does not reflect custom selection on the scatter plot
    screen shot 2018-10-31 at 7 04 15 pm

- [ ] I notice the color spectrum is gone in PDF
- [ ] Let's update the title to `Mutation Count vs Fraction of Genome Altered`
- [ ] Can we increase the chart height? 
     screen shot 2018-10-31 at 6 56 19 pm

File: src/pages/studyView/charts/ChartContainer.tsx [hunk 5]
```diff
@@ -353,42 +356,31 @@ export class ChartContainer extends React.Component<IChartContainerProps, {}> {
...
-                        width={400}
-                        height={380}
-                        onSelection={this.props.onValueSelection}
-                        data={this.props.promise.result}
-                        isLoading={this.props.selectedSamples.isPending}
-
-                        sampleToAnalysisGroup={this.props.sampleToAnalysisGroup!.result!}
-                        analysisGroups={this.props.analysisGroupsSettings.groups}
-                        analysisClinicalAttribute={this.props.analysisGroupsSettings.clinicalAttribute}
+                    <div style={{marginTop:-33}}>
```
Comment (zhx828) [inline src/pages/studyView/charts/ChartContainer.tsx]: What's this for?

File: src/shared/lib/onMobxPromise.ts [hunk 6]
```diff
@@ -1,16 +1,6 @@
 import {MobxPromise} from "mobxpromise";
 import {autorun, IReactionDisposer} from "mobx";
 
-export function onMobxPromise<T>(promise:MobxPromise<T>,
```
Comment (zhx828) [inline src/shared/lib/onMobxPromise.ts]: This is intensional?

Comment (adamabeshouse) [reply src/shared/lib/onMobxPromise.ts]: yes, theres actually no need for multiple signatures, and its causing problems so I just got rid of them

File: src/pages/studyView/charts/ChartContainer.tsx (hunk 5 above)
Comment (adamabeshouse) [reply src/pages/studyView/charts/ChartContainer.tsx]: this is to put less space between title and chart, I could not make the adjustment in victory so I just had to make the SVG bigger and then move it up

File: src/pages/studyView/charts/ChartContainer.tsx [hunk 7]
```diff
@@ -416,22 +408,26 @@ export class ChartContainer extends React.Component<IChartContainerProps, {}> {
...
-                    hideLabel={this.hideLabel}
-                    chartControls={this.chartControls}
-                    changeChartType={this.changeChartType}
-                    download={this.downloadTypes}
-                    setAnalysisGroups={this.setAnalysisGroups}
-                />
-                <StudyViewComponentLoader promises={this.loadingPromises}>
-                    {this.chart}
-                </StudyViewComponentLoader>
+                <span style={{position:"relative", zIndex:2}}>
```
[the remark this answers is not in the document]
Comment (adamabeshouse) [reply src/pages/studyView/charts/ChartContainer.tsx]: this is to make sure the chart header buttons go on top of the contents of the chart so they are clickable

**Pull request merged**